[REF] porting v8 base_delivery_...label and affiliate modules

bealdav opened the pull request:
Seems to be near to be ready for merge

bealdav committed:
[REF] remove stock.picking.in and .out

yvaucher commented:
You will need to rebase it your update of 8.0 branch has been done

bealdav committed:
[REF] port to v8 with old api

bealdav committed:
[PORT] delivery_carrier_deposit to v8, wip

bealdav committed:
[IMP] replace mako by qweb report in delivery_carrier_deposit

bealdav committed:
[MIG] to v8 base_deliv_label, deliv_carrier_deposit

yvaucher commented:
Some modules where moved from `__unported__` is that intentional ?

`__unported__/delivery_carrier_b2c/__openerp__.py → delivery_carrier_b2c/__openerp__.py`

bealdav commented:
@yvaucher  yes, just missed to change to installable: True for b2c. I'm working on Coliposte on V8

yvaucher commented:
Best practice is to do a PR per module port

Usually when I do this I have my 2 or 3 directory to work "à la Bazaar" and use symlink to not mix them

bealdav committed:
[FIX] installable to True

guewen commented on delivery_carrier_deposit/report/deposit_slip.xml:
```diff
@@ -0,0 +1,60 @@
+<openerp>
+<data>
+
+<template id="report_generic_deposit_slip">
```
I am not sure this report will be translated, I thought we have to put something like: https://github.com/odoo/odoo/blob/8.0/addons/account/views/report_invoice.xml#L149

bealdav replied:
fixed in th end of the file

guewen replied:
This is not fixed. Please see https://github.com/OCA/carrier-delivery/pull/33/files#r20997336

alexis-via replied:
@guewen About the language of the qweb report, by default, if you don't have the block of XML code like https://github.com/odoo/odoo/blob/8.0/addons/account/views/report_invoice.xml#L149 then the report is generated in the language of the user, so if the user is not using English, then the report will be translated.

guewen replied:
@alexis-via I know... I asked @bealdav to add it. But please read my comments **below** (https://github.com/OCA/carrier-delivery/pull/33/files#r20997336)

```
+<report id="report_deposit_slip"
+ model="deposit.slip"
+ string="Generic Deposit Slip"
+ report_type="qweb-pdf"
+ name="delivery_carrier_deposit.report_generic_deposit_slip"
+ file="delivery_carrier_deposit.report.deposit_slip"
+ />
```

It points to the `delivery_carrier_deposit.report_generic_deposit_slip` template, it must points to the translated template! 

```
+<report id="report_deposit_slip"
+ model="deposit.slip"
+ string="Generic Deposit Slip"
+ report_type="qweb-pdf"
+ name="delivery_carrier_deposit.report_generic_deposit_slip_trans"
+ file="delivery_carrier_deposit.report.deposit_slip"
+ />
```

guewen replied:
Or you consider the deposit slip should be in the language of the user, in that case you remove the `report_generic_deposit_slip_trans` template.

yvaucher replied:
This is fixed

guewen commented on delivery_carrier_deposit/report/deposit_slip.xml:
```diff
@@ -0,0 +1,60 @@
+    <thead>
+        <tr>
+            <th>Sender Ref</th>
+            <th>Name and address<br />destinataire</th>
```
destinataire → recipient?

guewen commented on delivery_carrier_deposit/report/deposit_slip.xml:
```diff
@@ -0,0 +1,60 @@
+
+<div class="page">
+
+<t t-foreach="docs" t-as="o">
```
Shouldn't be the loop outside of the `<div class="page">` element?

guewen commented on delivery_carrier_deposit/report/report.xml:
```diff
@@ -0,0 +1,27 @@
+
+<report id="report_deposit_slip"
+    model="deposit.slip"
+    string="Generic Deposit Slip pdf"
```
remove `pdf` from string

guewen commented on delivery_carrier_deposit/report/report.xml:
```diff
@@ -0,0 +1,27 @@
+    file="delivery_carrier_deposit.report.deposit_slip"
+    />
+
+<report id="report_deposit_slip_html"
```
Is the html report useful?

guewen commented:
Thanks for your work to port this module. A few things to address but seems rather good otherwise.

bealdav committed:
[IMP] add parcel_tracking field

bealdav committed:
[FIX] report deposit report

bealdav committed:
[FIX] upd translat pot

yvaucher commented on delivery_carrier_deposit/report/deposit_slip.xml:
```diff
@@ -0,0 +1,70 @@
+            <th>Parcel numbers</th>
+            <th class="small">ZIP</th>
+            <th class="small">Country</th>
+            <th class="small">Weight<br />(KG)</th>
```
kg instead of KG

kilo and grams must not be capital letters.

yvaucher commented on delivery_carrier_deposit/i18n/delivery_carrier_deposit.pot:
```diff
@@ -17,7 +17,17 @@ msgstr ""
+
+#. module: delivery_carrier_deposit
+#: view:website:delivery_carrier_deposit.report_generic_deposit_slip
+msgid "(KG)"
```
kg instead of KG
http://lamar.colostate.edu/~hillger/correct.htm

yvaucher commented on delivery_carrier_deposit/report/deposit_slip.xml:
```diff
@@ -0,0 +1,70 @@
+
+<h1>Deposit Slip</h1>
+
+    <div>DEPOSIT SLIP  <span t-field="o.carrier_type"/></div>
```
Are those Captal letters there for a reason ?

yvaucher commented on delivery_carrier_deposit/report/deposit_slip.xml:
```diff
@@ -0,0 +1,70 @@
+<h1>Deposit Slip</h1>
+
+    <div>DEPOSIT SLIP  <span t-field="o.carrier_type"/></div>
+    <div>N° CLIENT: </div>
```
N° is french

Here we want something like Customer No.

bealdav committed:
[FIX] from update main branch

yvaucher commented on delivery_carrier_deposit/report/deposit_slip.xml:
```diff
@@ -0,0 +1,70 @@
+    <div>DEPOSIT SLIP  <span t-field="o.carrier_type"/></div>
+    <div>N° CLIENT: </div>
+    <div>ACCOUNT NAME:  <span t-field="user.company_id.name"/></div>
+    <div>DOCUMENT N°:  <span t-field="o.name"/> - <span t-field="o.create_date"/></div>
```
Again, N° should be replaced by No. or no.

bealdav committed:
[FIX] report deposit_slip.xml

bealdav committed:
[IMP] upd pot

bealdav commented:
failed because depends of this PR https://github.com/OCA/partner-contact/pull/54

yvaucher commented:
Thanks for the changes

guewen commented on delivery_carrier_deposit/report/deposit_slip.xml:
```diff
@@ -0,0 +1,70 @@
+
+</template>
+
+<template id="report_eneric_deposit_slip_trans">
```
s/report_eneric_deposit_slip_trans/report_generic_deposit_slip_trans/

guewen commented on delivery_carrier_deposit/report/report.xml:
```diff
@@ -0,0 +1,19 @@
+    model="deposit.slip"
+    string="Generic Deposit Slip"
+    report_type="qweb-pdf"
+    name="delivery_carrier_deposit.report_generic_deposit_slip"
```
So you want the translated report here

guewen replied:
You are calling the inner part of the report, you need to call `report_generic_deposit_slip_trans`

guewen replied:
Still needs to be corrected here.

bealdav committed:
[FIX] call to report translation

bealdav committed:
[IMP] add _get_packages_from_picking()

bealdav commented:
I have added _get_packages_from_picking() method, 
I also done _get_packages_from_moves() that I can add. Should I also add this method ?

guewen commented:
> I have added _get_packages_from_picking() method,

What is it used for?

bealdav commented:
It's a common method which can be used for each carriers. It avoids to duplicate code ?

bealdav committed:
add carrier_id in tree view

bealdav committed:
[FIX] add explicitely attachment_id (previously included by orm)

alexis-via committed:
[FIX] add explicitely attachment_id (previously included by orm)

alexis-via committed:
[FIX] complete view

Commit:
Merge branch '8-base-carr' of https://github.com/bealdav/carrier-delivery into 8-base-carr

Commit:
delivery_carrier_deposit: port to new API
Add computed fields for total weight and total number of packages
Small enhancements in view and usability
Small cleanups

Commit:
Add XML header

bealdav committed:
[FIX] weight field type

Commit:
Add weight in form view of stock.quant.package
Better layout of default report for deposit slip
For stock.quant.package, order by id desc to have the latest first
Code cleanup

guewen commented:
Is it normal that delivery_carrier_deposit is now in the same PR?

guewen commented:
I work on this migration: https://github.com/OCA/carrier-delivery/pull/41 so I'm doing the review at the same time

alexis-via commented:
@guewen I ported the module delivery_carrier_deposit to the new API because I wanted to enhance this module for a customer project that is about to go live on v8 and I took the advantage of this to port it to the new API.

We are lacking time to split everything in multiple PRs, but the advantage is that we benefit from real-life testing on real-life projects... and it' the best way to find bugs and issues !

Commit:
delivery.py, migrate to new API

Commit:
Migration of carrier_account.py to new API

Commit:
Remove an unused file 'reports.xml'

Commit:
Button "Create Shipping Label" should be used when the picking in Done, because in v8 the packaging is handled in the "Transfer" wizard (in v7, the packaging could be made when the picking is waiting/ready)
Display number of packages on deposit slip
Add weight in tree view of packages

Commit:
Migration to new api: stock.py and some fixes

guewen commented on base_delivery_carrier_label/stock.py:
```diff
@@ -23,9 +23,23 @@
-class stock_picking(orm.Model):
+class StockQuantPackage(orm.Model):
+    _inherit = 'stock.quant.package'
+    _order = 'id desc'
```
What is the reason for overriding this? The default is:

```
    _parent_order = 'name'
    _order = 'parent_left'

```

guewen commented on delivery_carrier_deposit/stock.py:
```diff
@@ -0,0 +1,127 @@
+
+    name = fields.Char(
+        string='Name', readonly=True, states={'draft': [('readonly', False)]},
+        default='/', copy='/')
```
`copy` is supposed to be a boolean

guewen commented on delivery_carrier_deposit/stock.py:
```diff
@@ -0,0 +1,127 @@
+    def copy(self, default=None):
+        default = dict(default or {})
+        default['name'] = self.env['ir.sequence'].next_by_code(
+            'delivery.deposit')
```
Why not put `default['name'] = '/'`, so `create()` will call the next sequence number? (remove duplication)

guewen commented on delivery_carrier_deposit/wizard/deposit.py:
```diff
@@ -0,0 +1,81 @@
+    @api.multi
+    def create_deposit_slip(self):
+        # I can't set api.one because I return an action
+        assert len(self) == 1, 'only 1 recordset allowed'
```
Use `self.ensure_one()`

guewen commented on delivery_carrier_deposit/wizard/deposit.py:
```diff
@@ -0,0 +1,81 @@
+    def create_deposit_slip(self):
+        # I can't set api.one because I return an action
+        assert len(self) == 1, 'only 1 recordset allowed'
+        self = self[0]
```
You can remove this line. When a recordset contains 1 record, it is considered as a "singleton" and the attributes work directly on the record.

guewen commented:
@bealdav can you review and merge that in your PR please? https://github.com/bealdav/carrier-delivery/pull/1

guewen commented:
We have currently no way to print the labels from the new javascript "barcode interface". Are you going to work on that? I am not sure how we should do, because the picking disappears once the packing is finished. Do you have an idea?

bealdav commented:
@guewen no idea for barcode interface. We do not use this interface for now. Alexis tried but js seems to be less inefficient than python.
As you can see here https://github.com/akretion/carrier-delivery-colipostefr/blob/8.0/delivery_carrier_label_colissimo/stock.py#L19, you have to override 2 methods if you want to control the 2 interfaces: maybe to switch in base module

bealdav commented:
@guewen I'll review and test your works early next week on a real cases. Thanks a lot frr your work

guewen commented:
> As you can see here https://github.com/akretion/carrier-delivery-colipostefr/blob/8.0/delivery_carrier_label_colissimo/stock.py#L19, you have to override 2 methods if you want to control the 2 interfaces: maybe to switch in base module

Thanks. The issue here won't be to generate the labels but to allow the user to print them :unamused:

Commit:
Typo in string

Commit:
Remove the selection wrappers

Use the name of the method as string instead

Commit:
Shadowing function name

Commit:
Wrong model for carrier

Commit:
Wrong usage of res_partner.address_get()

Commit:
Return a browse_record

bealdav commented:
LGTM for me

guewen commented on base_delivery_carrier_label/stock.py:
```diff
@@ -0,0 +1,298 @@
+
+class StockQuantPackage(models.Model):
+    _inherit = 'stock.quant.package'
+    _order = 'id desc'
```
I already asked this question and think I got not answer.
The override of the order here is highly suspicious.

For the record, here is the one defined in `stock`:

```
 _parent_order = 'name'
_order = 'parent_left'
```

The order here is based on `parent_left` so I guess it is for respecting the tree's order.
Maybe you want to change the `_parent_order`, but what is the reason at a first place to change that in this module?

:-1: for that override

alexis-via replied:
Well, I just wanted to see the latest packages at the top, and not the oldest packages at the top. I'll revert it if you're not happy with this ; it's not important, it's just a small usabilty stuff than I can put in another module.

Commit:
Remove _order on stock.quant.packages, as asked by Guewen

bealdav committed:
[FIX] copy in deposit.slip

bealdav committed:
Merge branch '8-base-carr' of github.com:bealdav/carrier-delivery into 8-base-carr

bealdav committed:
[FIX] writing package_id field in shipping.label

Commit:
[Usability] tree view of deposity.slip: blue when draft

Commit:
[Usability] Better name_get on stock.quant.package

guewen commented:
@bealdav is it still work in progress?

bealdav commented:
I have not fixed the translation problem in the report report_deposit_slip. I you have some time for this thank you.

alexis-via commented:
For me this is stable and can be merged. I already use this branch in production.

yvaucher commented:
I put it as needs review then

pedrobaeza commented:
Please put module descriptions with OCA template and on a README.rst, and I will merge this.

yvaucher commented:
Almost there, as @pedrobaeza said please move description in README.rst using template:
https://github.com/OCA/maintainers-tools/blob/master/template/module/README.rst

guewen commented on base_delivery_carrier_label/stock.py:
```diff
@@ -0,0 +1,307 @@
+        operations = operation_obj.search(
+            [('result_package_id', '!=', False),
+             ('picking_id', '=', self.id)]
+        )
```
I think I spotted an issue there but I am not sure whether it is a misuse from my part or... I dont know.

It won't work in this scenario:
- Configure the warehouse in 3 steps (pick + pack + ship)
- On the pick, put all the lines in packages
- On the pack, keep the lines in the same packages
- On the ship, keep the lines in the same packages again

The packages have no `result_package_id`, so no labels are generated.

@bealdav @alexis-via any idea?

guewen replied:
> The packages have no result_package_id, so no labels are generated.

If I guess (can't say "understand" at this point) correctly, the `stock.pack.operation` could have a `package_id`, a `result_package_id` or both. If we have only a `package_id`, we should keep this package. So, the logic should be changed throughout the module to look first in `result_package_id` then in `package_id`.
Does this analysis sounds good to you?

guewen commented:
Similar scenario that don't work as we could expect not strictly related to this module I fear (in fact I observed it in delivery_carrier_colipostefr but I wonder if it is not a bug in the stock module):
-  Configure the warehouse in 3 steps (pick + pack + ship)
-  On the pick, put all the lines in packages
-  On the pack, keep the same packages
-  On the ship, transfer the existing packages in new packages

The packages in ship have no product_id so generation of labels might fails.
Example, here https://github.com/akretion/carrier-delivery-colipostefr/blob/8.0/delivery_carrier_colipostefr/stock.py#L236 the weight is 0 because `pack_op.product_id` is empty.

Commit:
Move module description in README.rst

bealdav commented:
@guewen, I don't tested it with these scenari, then I suppose your are right. I need some time for it.
Thanks to point it

guewen commented:
> @guewen, I don't tested it with these scenari, then I suppose your are right. I need some time for it.
> Thanks to point it

I will try that and make a new PR upon this one for the first scenario.
The second one I am unsure about how it should be resolved.

Commit:
Allow to generate labels when we moved an entire package

Take the destination package. If empty, the package is moved so take the source
one.

guewen commented:
Here it is: https://github.com/bealdav/carrier-delivery/issues/2

guewen commented:
I can't reproduce the scenario in https://github.com/OCA/carrier-delivery/pull/33#issuecomment-72036213 so it was maybe related to the fix I did in bealdav#2 and https://github.com/akretion/carrier-delivery-colipostefr/pull/5

UPDATE: yes I can reproduce it, working on a fix.
UPDATE: got it in https://github.com/akretion/carrier-delivery-colipostefr/pull/5

Commit:
The options must be filtered by the selected carrier

bealdav committed:
Merge pull request #2 from guewen/8-base-carr_result_package_id

Allow to generate labels when we moved an entire package

yvaucher commented:
@guewen @bealdav can we remove the needs fixing here?

yvaucher commented:
:+1: 

Thanks for the work @bealdav I'll proceed with the merge of this PR
